### PR TITLE
Add integration test for Logcollector: invalid localfile configuration

### LINKED ISSUE #2391: Logcollector - Test Wazuh behavior's when setting an invalid <localfile>  configuration.
Installation | Component | Version | Related issue
-- | -- | -- | --
Manager/Agent | Logcollector | 4.3.0 | wazuh/wazuh#6463

# Objective

1. Create a test that checks the manager or agent behavior when they have an invalid ```<localfile>```  configuration.

# Scope

**"Logcollector" and "Agentd" must be tested on the Manager and the Agent, respectively.**

## Logcollector (Manager)

- [x] CentOS
- [x] Ubuntu

## Agentd (Agent)

- [x] Windows
- [x] CentOS
- [x] Ubuntu
- [x] MacOS
- [x] Solaris

# Context

The Logcollector's **configuration parser** reads each ```<localfile>``` stanza in ```ossec.conf``` and ```agent.conf```. After that, it performs some extra checks:

1. On Windows, ```<location>``` must be either "Application", "System" or "Security" when ```<log_format>``` is "eventchannel".
2. Expand ```<location>``` if it is a glob pattern.
3. ```<log_format>``` and ```<location>``` must exist.

> The **configuration parser** used to check that ```<location>``` exists after checking points 1 and 2, throwing a memory error if ```<location>``` was missing.

# Tests

- [x] Given that the `location` option is not present in the `localfile` section in the `ossec.conf` file. When the `manager` is restarted, the `manager` refuses to restart and the expected messages are in the log file.
- [x] Given that the `log_format` option is not present in the `localfile` section in the `ossec.conf` file. When the `manager` is restarted, the `manager` refuses to restart and the expected messages are in the log file.
- [x] Given that the `location` option is not present in the `localfile` section in the `ossec.conf` file. When the `agent` is restarted, the `agent` refuses to restart and the expected messages are in the log file.
- [x] Given that the `log_format` option is not present in the `localfile` section in the `ossec.conf` file. When the `agent` is restarted, the `agent` refuses to restart and the expected messages are in the log file.
- [x] Given that the `location` option is not present in the `localfile` section in the `agent.conf` file. When the `agent` is restarted, the `agent` continues running and the expected messages are in the log file.
- [x] Given that the `log_format` option is not present in the `localfile` section in the `agent.conf` file. When the `agent` is restarted, the `agent` continues running and the expected messages are in the log file.

### Expected result when trying to restart Wazuh after an invalid config

```
2021/12/22 16:14:47 wazuh-logcollector: ERROR: (1902): Missing 'location' element.
2021/12/22 16:14:47 wazuh-logcollector: ERROR: (1202): Configuration error at 'etc/ossec.conf'.
2021/12/22 16:14:47 wazuh-logcollector: CRITICAL: (1202): Configuration error at 'etc/ossec.conf'.
wazuh-logcollector: Configuration error. Exiting
```

### PR DESCRIPTION
**Pull request by danisan90**
|Related issue|
|---|
| #2391 |

## Description
This PR closes #2391 

This PR adds a test that checks the manager or agent behavior when they have an invalid <localfile> configuration. It changes the `ossec.conf` or the `agent.conf` with a configuration without the `<location>` generating expected error messages. To pass the test this are the expected results:
- wazuh-manager:
```
2021/12/22 16:14:47 wazuh-logcollector: ERROR: (1902): Missing 'location' element.
2021/12/22 16:14:47 wazuh-logcollector: ERROR: (1202): Configuration error at 'etc/ossec.conf'.
2021/12/22 16:14:47 wazuh-logcollector: CRITICAL: (1202): Configuration error at 'etc/ossec.conf'.
wazuh-logcollector: Configuration error. Exiting
```
- wazuh-agent
```
ERROR: (1902): Missing 'location' element.
ERROR: (1202): Configuration error at '/var/ossec/etc/shared/agent.conf'.
```

### **Packages**
 System|  Version | Link |
|---   |---    |---    |
| Linux Agent|   4.3.0    | https://packages-dev.wazuh.com/warehouse/test/4.3/rpm/var/wazuh-agent-4.3.0-dani.logcollector.2391.el5.x86_64.rpm |
| Linux Manager|   4.3.0    | https://packages-dev.wazuh.com/warehouse/test/4.3/rpm/var/wazuh-manager-4.3.0-dani.logcollector.2391.x86_64.rpm |
| Windows Agent|   4.3.0    | https://packages-dev.wazuh.com/warehouse/test/4.3/windows/wazuh-agent-4.3.0-dani.logcollector.2391.msi|
| solarisOS Agent|   4.3.0    | https://packages-dev.wazuh.com/warehouse/test/4.3/solaris/i386/11/wazuh-agent_v4.3.0-dani.logcollector.2391-sol11-i386.p5p |
| macOS Agent|   4.3.0    | https://packages-dev.wazuh.com/warehouse/test/4.3/macos/wazuh-agent-4.3.0-dani.logcollector.2391.pkg|

#### Reports
| OS | Type | Results | Date |
|--|--|--|--|
|Centos|Manager/Local| [🟢 ](https://github.com/wazuh/wazuh-qa/files/7916334/2391ManagerCentos.zip)| 01/21/2022
|Centos|Agent/Local| [🟢 ](https://github.com/wazuh/wazuh-qa/files/7916338/2391AgentCentos.zip)| 01/21/2022
|Solaris|Agent/ Local| [🟢 ](https://github.com/wazuh/wazuh-qa/files/7926520/2391-Agent-SolarisOS.zip)| 01/24/2022
|Windows|Agent/ Local| [🟢 ](https://github.com/wazuh/wazuh-qa/files/7927909/2391-Agent-Windows.zip)| 01/24/2022
|MacOS|Agent/ Local|[🟢 ](https://ci.wazuh.info/view/Tests/job/Test_integration/19240/)| 01/21/2022
|Centos|Manager/Jenkins| [🔴  ](https://ci.wazuh.info/view/Tests/job/Test_integration/19308/console)| 01/21/2022
|Centos|Agent/Jenkins| [🟢 ](https://ci.wazuh.info/view/Tests/job/Test_integration/19301/consoleFull)| 01/24/2022
|Windows|Agent/Jenkins| [🟢 ](https://ci.wazuh.info/view/Tests/job/Test_integration/19300/consoleFull)| 01/24/2022
|Solaris|Agent/ Jenkins| [🟢 ](https://ci.wazuh.info/view/Tests/job/Test_integration/19229/)| 01/21/2022
|MacOS|Agent/ Jenkins|[🟢 ](https://github.com/wazuh/wazuh-qa/files/7934158/2391AgentmacOS.zip)| 01/25/2022

## QA-Docs execution
| Status | Output |
| --- | --- |
| 🟢 | [qa-docOutput.zip](https://github.com/wazuh/wazuh-qa/files/7936043/qa-docOutput.zip) |

- [x] Proven that tests **pass** when they have to pass.
- [x] Proven that tests **fail** when they have to fail.

- [x] Python codebase satisfies PEP-8 style style guide. `pycodestyle --max-line-length=120 --show-source --show-pep8 file.py`.
- [x] Python codebase is documented following the Google Style for Python docstrings.
- [x] The test is documented in wazuh-qa/docs.

### CODE REVIEW

**Comment by mauromalara** [comment]
# Test case design
It is necessary to add the following test cases:

1. The `location` option is not present in the `localfile` section in the `ossec.conf` file (MANAGER)
2. The `log_format` option is not present in the `localfile` section in the `ossec.conf` file (MANAGER)
3. The `location` option is not present in the `localfile` section in the `ossec.conf` file (AGENT)
4. The `log_format` option is not present in the `localfile` section in the `ossec.conf` file (AGENT)
5. The `location` option is not present in the `localfile` section in the `agent.conf` file (AGENT)
6. The `log_format` option is not present in the `localfile` section in the `agent.conf` file (AGENT)

File: tests/integration/test_logcollector/test_localfile/test_invalid_agent_localfile_config.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+]
+
+cases = [
+    {
+        'params': {
+            'LOCATION': os.path.join(temp_dir, 'wazuh-testing', files[0]),
+            'LOG_FORMAT': None
+        },
+        'metadata': {
+            'regex': callback_configuration_error
```
Comment (Rebits) [inline tests/integration/test_logcollector/test_localfile/test_invalid_agent_localfile_config.py:120]: Why there is a metadata var with the same callback for the two test cases?

File: tests/integration/test_logcollector/test_localfile/test_invalid_agent_localfile_config.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    input_description: A YAML file with the invalid configurations.
+
+    expected_output:
+        - 'Did not receive expected "ERROR: ...: Configuration error at event'
+
+    tags:
+        - logcollector
+    '''
+    metadata = get_configuration.get('metadata')
+    wazuh_log_monitor.start(timeout=5, callback=metadata['regex'],
```
Comment (Rebits) [inline tests/integration/test_logcollector/test_localfile/test_invalid_agent_localfile_config.py:191]: Rename the `regex` metadata for `callback`
```suggestion
    wazuh_log_monitor.start(timeout=5, callback=metadata['callback'],
```

File: tests/integration/test_logcollector/test_localfile/test_invalid_wazuh_conf.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    {
+        'params': {
+            'LOCATION': os.path.join(temp_dir, 'wazuh-testing', files[0]),
+            'LOG_FORMAT': None
+        },
+        'metadata': {
+            'regex': callback_configuration_error
+        }
+    }
+]
```
Comment (Rebits) [inline tests/integration/test_logcollector/test_localfile/test_invalid_wazuh_conf.py:135]: Same as before

**Review by Rebits — CHANGES_REQUESTED** [review]
GJ, but some changes are required

File: tests/integration/test_logcollector/test_localfile/test_invalid_agent_localfile_config.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+pytestmark = [pytest.mark.tier(level=0), pytest.mark.agent]
+
+# Variables
+wazuh_log_monitor = FileMonitor(LOG_FILE_PATH)
+files = ['test.txt']
+
+# Configuration
+daemons_handler_configuration = {
+    'daemons': [LOGCOLLECTOR_DAEMON],
+    'ignore_errors': True
```
Comment (Rebits) [inline tests/integration/test_logcollector/test_localfile/test_invalid_agent_localfile_config.py:102]: Using this structure is not possible to check if adding invalid configuration in `agent.conf` the agent still running.

**Comment by mauromalara** [comment]
# 17/03/2022
## Packages
| Version | Revision | Repository |
|---|---|---|
| 4.3.0 | 1 | pre-release |

## Test execution

| OS | Target | Result |
|--- |---     |---     |
| CentOS | Manager | [🟢](https://ci.wazuh.info/job/Test_integration/23958/artifact/manager_tests_Test_integration_B23958_20220318131450.txt) - [Build](https://ci.wazuh.info/job/Test_integration/23958/) |
| Ubuntu | Manager | [🟢](https://ci.wazuh.info/job/Test_integration/23899/artifact/manager_tests_Test_integration_B23899_20220318015741.txt) - [Build](https://ci.wazuh.info/job/Test_integration/23899/) |
| CentOS | Agent   | [🟢](https://ci.wazuh.info/job/Test_integration/23897/artifact/centos_agent_tests_Test_integration_B23897_20220318015737.txt) - [Build](https://ci.wazuh.info/job/Test_integration/23897/) |
| Ubuntu | Agent   | [🟢](https://ci.wazuh.info/job/Test_integration/23897/artifact/ubuntu_agent_tests_Test_integration_B23897_20220318015737.txt) - [Build](https://ci.wazuh.info/job/Test_integration/23897/) |
| Windows| Agent   | [🟢](https://ci.wazuh.info/job/Test_integration/23897/artifact/windows_agent_tests_Test_integration_B23897_20220318015737.txt) - [Build](https://ci.wazuh.info/job/Test_integration/23897/) |
| Solaris| Agent   | [🟢](https://ci.wazuh.info/job/Test_integration/23897/artifact/solaris_agent_tests_Test_integration_B23897_20220318015737.txt) - [Build](https://ci.wazuh.info/job/Test_integration/23897/) |
| MacOS  | Agent   | [🟢](https://ci.wazuh.info/job/Test_integration/23897/artifact/macos_agent_tests_Test_integration_B23897_20220318015737.txt) - [Build](https://ci.wazuh.info/job/Test_integration/23897/) |

#### References

|Color|Status |
|:--:|:--|
|🟢|All tests passed successfully|
|🟡|All tests passed but there are some warnings|
|🔴|Some tests have failures or errors|
|🔵|Test execution in progress|
|⚫|To Do|
|🟠|Jenkins provision fails|

**Comment by mauromalara** [comment]
## 18-03-2022
### Package
| Version | Revision | Repository |
|---|---|---|
|4.3.0| 1 | pre-release |


## Testing

### Logcollector

|  OS  | Jenkins     | Notes
|---    |---    |---    |
| R1   | [🔵](link-to-the-report-r1) - [Build](https://ci.wazuh.info/job/Test_integration_launcher/39/)   | |
| R2   | [🔵](link-to-the-report-r1) - [Build](https://ci.wazuh.info/job/Test_integration_launcher/40/)   | |
| R3   | [🔵](link-to-the-report-r1) - [Build](https://ci.wazuh.info/job/Test_integration_launcher/41/)   | |

* * * 

- :green_circle:: All pass
- :yellow_circle:: Some warnings
- :red_circle:: Some errors/fails
- :large_blue_circle:: In progress

**Review by Rebits — DISMISSED** [review]
LGTM

**Comment by Rebits** [comment]
## 05/042022

## Testing

### `test_logcollector`
|  OS  |  Jenkins     | Notes
|---   |---    |---    |
| R1   | [:green_circle: ](https://ci.wazuh.info/job/Test_integration/24703/ )   | |

* * * 

- :green_circle:: All pass
- :yellow_circle:: Some warnings
- :red_circle:: Some errors/fails
- :large_blue_circle:: In progress

**Review by Rebits — DISMISSED** [review]

**Review by jmv74211 — APPROVED** [review]

**Review by Rebits — DISMISSED** [review]

**Comment by mauromalara** [comment]
## 06/04/2022
### Package
| Version | Revision | Link|
|---|---|---|
|4.3.0|1|pre-release|


## Testing

### Module 1

|  Target  | Jenkins     | Notes
|---    |---    |---    |
| Agent   | [:large_blue_circle:]() - [Build](https://ci.wazuh.info/job/Test_integration/24784/)   | - |
| Manager   | [:large_blue_circle:]() - [Build](https://ci.wazuh.info/job/Test_integration/24785/)   | - |

* * *

:large_blue_circle:: The tests are being executed.
:green_circle:: All tests passed successfully.
:yellow_circle:: Some tests passed with warnings.
:orange_circle:: Some errors or failures of Jenkins.
:red_circle:: Some tests failed or had errors.

**Review by Rebits — APPROVED** [review]

**Pull request merged**